### ISSUE TITLE
GKE: Stacks and CR of stack got deleted for no reason

### ISSUE THREAD

**Issue by turkenh** [crossplaneio/crossplane#895]
### What happened?
On a GKE cluster (v1.15.3-gke.1), I installed crossplane (v0.3.0) and deployed wordpress by following the stacks guide (for GCP). Everything was fine and I was able to reach wordpress app from browser. Then, I experimented a bit, several `get's` on different objects, which I expect to be unrelated.

After about 21h, I checked the cluster again but my wordpressinstance seems to be gone. Digging further, I realized that, the stacks (gcp and wordpress) already gone.

```
$ kubectl get stacks --all-namespaces
No resources found
$ kubectl get stackinstalls --all-namespaces
NAMESPACE          NAME                     READY   SOURCE   PACKAGE                                    CRD   AGE
app-project1-dev   sample-stack-wordpress   True             crossplane/sample-stack-wordpress:latest         20h
$ kubectl get clusterstackinstalls --all-namespaces
NAMESPACE   NAME        READY   SOURCE   PACKAGE                       CRD   AGE
gcp         stack-gcp   True             crossplane/stack-gcp:master         21h
```

Most of the CRDs belonging to stacks also gone but not all:

```
$ kubectl get crds | grep gcp
cloudsqlinstances.database.gcp.crossplane.io            2019-10-06T10:41:37Z
subnetworks.compute.gcp.crossplane.io                   2019-10-06T10:41:37Z
```

When I check the logs of stack-manager, I see some connectivity issues at some point in time but finally recovered which could be related to the issue.

```
2019-10-07T02:40:47.548Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "stackinstall.stacks.crossplane.io", "request": "app-project1-dev/sample-stack-wordpress"}
2019-10-07T02:40:47.553Z	DEBUG	crossplane.stackinstall.stacks.crossplane.io	reconciling	{"kind": "stacks.crossplane.io/v1alpha1, Kind=ClusterStackInstall", "request": "gcp/stack-gcp"}
2019-10-07T02:40:47.553Z	DEBUG	crossplane.util	getting pod	{"name": "crossplane-system/crossplane-stack-manager-7fd8d6b9fc-mspt7"}
2019-10-07T02:40:47.553Z	DEBUG	crossplane.stackinstall.stacks.crossplane.io	updating not supported yet	{"clusterstackinstall.stacks.crossplane.io/v1alpha1": "gcp/stack-gcp"}
2019-10-07T02:40:47.553Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "clusterstackinstall.stacks.crossplane.io", "request": "gcp/stack-gcp"}
E1007 02:49:27.141062       1 reflector.go:270] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to watch *v1alpha1.Stack: Get https://10.8.0.1:443/apis/stacks.crossplane.io/v1alpha1/stacks?resourceVersion=125697&timeoutSeconds=364&watch=true: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:49:27.142221       1 reflector.go:270] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:204: Failed to watch *v1.Pod: Get https://10.8.0.1:443/api/v1/pods?resourceVersion=180641&timeoutSeconds=386&watch=true: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:49:27.142289       1 reflector.go:270] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to watch *v1alpha1.ClusterStackInstall: Get https://10.8.0.1:443/apis/stacks.crossplane.io/v1alpha1/clusterstackinstalls?resourceVersion=1582&timeoutSeconds=423&watch=true: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:49:27.142340       1 reflector.go:270] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to watch *v1alpha1.StackInstall: Get https://10.8.0.1:443/apis/stacks.crossplane.io/v1alpha1/stackinstalls?resourceVersion=3464&timeoutSeconds=450&watch=true: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:49:27.142393       1 reflector.go:270] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:204: Failed to watch *v1.Job: Get https://10.8.0.1:443/apis/batch/v1/jobs?resourceVersion=125695&timeoutSeconds=397&watch=true: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:49:28.143938       1 reflector.go:126] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1alpha1.Stack: Get https://10.8.0.1:443/apis/stacks.crossplane.io/v1alpha1/stacks?limit=500&resourceVersion=0: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:49:28.145037       1 reflector.go:126] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:204: Failed to list *v1.Pod: Get https://10.8.0.1:443/api/v1/pods?limit=500&resourceVersion=0: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:49:28.146116       1 reflector.go:126] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1alpha1.ClusterStackInstall: Get https://10.8.0.1:443/apis/stacks.crossplane.io/v1alpha1/clusterstackinstalls?limit=500&resourceVersion=0: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:49:28.147583       1 reflector.go:126] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1alpha1.StackInstall: Get https://10.8.0.1:443/apis/stacks.crossplane.io/v1alpha1/stackinstalls?limit=500&resourceVersion=0: dial tcp 10.8.0.1:443: connect: connection refused
...
E1007 02:50:01.220220       1 reflector.go:126] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1alpha1.StackInstall: Get https://10.8.0.1:443/apis/stacks.crossplane.io/v1alpha1/stackinstalls?limit=500&resourceVersion=0: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:50:02.217546       1 reflector.go:126] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1alpha1.ClusterStackInstall: Get https://10.8.0.1:443/apis/stacks.crossplane.io/v1alpha1/clusterstackinstalls?limit=500&resourceVersion=0: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:50:02.218231       1 reflector.go:126] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1alpha1.Stack: Get https://10.8.0.1:443/apis/stacks.crossplane.io/v1alpha1/stacks?limit=500&resourceVersion=0: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:50:02.219803       1 reflector.go:126] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:204: Failed to list *v1.Job: Get https://10.8.0.1:443/apis/batch/v1/jobs?limit=500&resourceVersion=0: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:50:02.221845       1 reflector.go:126] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1alpha1.StackInstall: Get https://10.8.0.1:443/apis/stacks.crossplane.io/v1alpha1/stackinstalls?limit=500&resourceVersion=0: dial tcp 10.8.0.1:443: connect: connection refused
E1007 02:50:02.221959       1 reflector.go:126] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:204: Failed to list *v1.Pod: Get https://10.8.0.1:443/api/v1/pods?limit=500&resourceVersion=0: dial tcp 10.8.0.1:443: connect: connection refused
2019-10-07T02:50:12.659Z	DEBUG	crossplane.stackinstall.stacks.crossplane.io	reconciling	{"kind": "stacks.crossplane.io/v1alpha1, Kind=ClusterStackInstall", "request": "gcp/stack-gcp"}
2019-10-07T02:50:12.659Z	DEBUG	crossplane.util	getting pod	{"name": "crossplane-system/crossplane-stack-manager-7fd8d6b9fc-mspt7"}
2019-10-07T02:50:12.659Z	DEBUG	crossplane.stackinstall.stacks.crossplane.io	updating not supported yet	{"clusterstackinstall.stacks.crossplane.io/v1alpha1": "gcp/stack-gcp"}
2019-10-07T02:50:12.659Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "clusterstackinstall.stacks.crossplane.io", "request": "gcp/stack-gcp"}
2019-10-07T02:50:12.909Z	DEBUG	crossplane.stackinstall.stacks.crossplane.io	reconciling	{"kind": "stacks.crossplane.io/v1alpha1, Kind=StackInstall", "request": "app-project1-dev/sample-stack-wordpress"}
2019-10-07T02:50:12.909Z	DEBUG	crossplane.util	getting pod	{"name": "crossplane-system/crossplane-stack-manager-7fd8d6b9fc-mspt7"}
2019-10-07T02:50:12.909Z	DEBUG	crossplane.stackinstall.stacks.crossplane.io	updating not supported yet	{"stackinstall.stacks.crossplane.io/v1alpha1": "app-project1-dev/sample-stack-wordpress"}
2019-10-07T02:50:12.909Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "stackinstall.stacks.crossplane.io", "request": "app-project1-dev/sample-stack-wordpress"}
2019-10-07T03:50:12.659Z	DEBUG	crossplane.stackinstall.stacks.crossplane.io	reconciling	{"kind": "stacks.crossplane.io/v1alpha1, Kind=ClusterStackInstall", "request": "gcp/stack-gcp"}
2019-10-07T03:50:12.659Z	DEBUG	crossplane.util	getting pod	{"name": "crossplane-system/crossplane-stack-manager-7fd8d6b9fc-mspt7"}
2019-10-07T03:50:12.659Z	DEBUG	crossplane.stackinstall.stacks.crossplane.io	updating not supported yet	{"clusterstackinstall.stacks.crossplane.io/v1alpha1": "gcp/stack-gcp"}
2019-10-07T03:50:12.659Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "clusterstackinstall.stacks.crossplane.io", "request": "gcp/stack-gcp"}
2019-10-07T03:50:12.909Z	DEBUG	crossplane.stackinstall.stacks.crossplane.io	reconciling	{"kind": "stacks.crossplane.io/v1alpha1, Kind=StackInstall", "request": "app-project1-dev/sample-stack-wordpress"}
```

Some more info, which could help:
```
$ kubectl get mysqlinstances --all-namespaces
NAMESPACE          NAME                                                   STATUS   CLASS            VERSION   AGE
app-project1-dev   wordpress-mysql-09210911-2260-4106-a01c-9b860aefa24b   Bound    standard-mysql   5.7       21h
$ kubectl get cloudsqlinstances --all-namespaces
NAMESPACE   NAME                                                 STATUS   STATE   CLASS               VERSION     AGE
gcp         mysqlinstance-e55cd378-4009-4274-92ee-43b4f268154e                    standard-cloudsql   MYSQL_5_7   11h
$ kubectl get kubernetesclusters --all-namespaces
NAMESPACE          NAME                                                     STATUS   CLUSTER-CLASS      CLUSTER-REF   AGE
app-project1-dev   wordpress-cluster-09210911-2260-4106-a01c-9b860aefa24b   Bound    standard-cluster                 21h
$ kubectl get gkecluster --all-namespaces
error: the server doesn't have a resource type "gkecluster"
$ kubectl get kubernetesapplications --all-namespaces
No resources found
$ kubectl get kubernetesapplicationresources --all-namespaces
NAMESPACE          NAME                                                             TEMPLATE-KIND   TEMPLATE-NAME   CLUSTER                                                  STATUS
app-project1-dev   wordpress-demo-deployment-09210911-2260-4106-a01c-9b860aefa24b   Deployment      wordpress       wordpress-cluster-09210911-2260-4106-a01c-9b860aefa24b   Submitted
app-project1-dev   wordpress-demo-namespace-09210911-2260-4106-a01c-9b860aefa24b    Namespace       wordpress       wordpress-cluster-09210911-2260-4106-a01c-9b860aefa24b   Submitted
app-project1-dev   wordpress-demo-service-09210911-2260-4106-a01c-9b860aefa24b      Service         wordpress       wordpress-cluster-09210911-2260-4106-a01c-9b860aefa24b   Submitted
$ kubectl get wordpressinstances --all-namespaces
error: the server doesn't have a resource type "wordpressinstances"
```
Restarting stack manager pod didn't help.

### How can we reproduce it?

Follow stacks guide, leave the cluster. If the issue is related to aforementioned connectivity issue, breaking connectivity betweek stack manager and api server could help to reproduce.

### What environment did it happen in?
Crossplane version: v0.3.0

**Comment by turkenh** [comment]
Seems to be related to https://github.com/crossplaneio/crossplane/issues/854 at first look. 
But what I understand especially from [this comment](https://github.com/crossplaneio/crossplane/issues/854#issuecomment-536136561) which means aws stack is still there. On my side, I could neither see GCP stack nor any pods in `gcp` namespace .
Still, the similarity between the two issues (at different controller) could signal some problem related to the runtime they both are relying.

**Comment by jbw976** [comment]
This is definitely a concerning issue @turkenh, thank you for capturing it!  I'm surprised this is happening on a steady/consistent cluster like GKE instead of a local laptop dev cluster like KinD/Minikube that gets put to sleep.

It sounds like there `Stack` object is getting garbage collected or deleted for some reason and then that is causing a whole cascading set of deletes because it is set as the owner for a lot of other resources (CRDs owned by the stack, `Deployment` to run the controller of the stack, etc.).

If we can figure out why the `Stack` record is getting deleted (or garbage collected), then that would probably answer why everything else is getting deleted.

**Comment by turkenh** [comment]
TL; DR, It looks like we are using **Cross-namespace owner references** which are said to be disallowed by design in [documentation](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents), which results crossplane resources to get garbage collected.

> Note: Cross-namespace owner references are disallowed by design. This means: 1) Namespace-scoped dependents can only specify owners in the same namespace, and owners that are cluster-scoped. 2) Cluster-scoped dependents can only specify cluster-scoped owners, but not namespace-scoped owners.

Here are the logs of api-server for initial **delete** requests:
```
A 2019-10-06T20:55:38.007634Z Kubernetes Apiservice Requests delete gcp:kubernetescluster-98680e35-881f-4dc2-a0de-2e2ad828f201 system:serviceaccount:kube-system:generic-garbage-collector {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:kube-system:generic-garbage-collector"},"authorizationInfo":[{"granted":true,"permission":"io.crossplane.gcp.compute.v1alpha2.gkeclusters.delete","resource":"compute.gcp.crossplan… Kubernetes Apiservice Requests delete gcp:kubernetescluster-98680e35-881f-4dc2-a0de-2e2ad828f201 system:serviceaccount:kube-system:generic-garbage-collector 
A 2019-10-06T20:55:38.031218Z Kubernetes Apiservice Requests delete gcp:mysqlinstance-e55cd378-4009-4274-92ee-43b4f268154e system:serviceaccount:kube-system:generic-garbage-collector {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:kube-system:generic-garbage-collector"},"authorizationInfo":[{"granted":true,"permission":"io.crossplane.gcp.database.v1alpha2.cloudsqlinstances.delete","resource":"database.gcp.c… Kubernetes Apiservice Requests delete gcp:mysqlinstance-e55cd378-4009-4274-92ee-43b4f268154e system:serviceaccount:kube-system:generic-garbage-collector 
A 2019-10-06T20:55:38.628655Z Kubernetes Apiservice Requests delete gcp:stack-gcp system:serviceaccount:kube-system:generic-garbage-collector {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:kube-system:generic-garbage-collector"},"authorizationInfo":[{"granted":true,"permission":"io.k8s.batch.v1.jobs.delete","resource":"batch/v1/namespaces/gcp/jobs/stack-gcp"}],"meth… Kubernetes Apiservice Requests delete gcp:stack-gcp system:serviceaccount:kube-system:generic-garbage-collector 
A 2019-10-06T20:55:38.919874Z Kubernetes Apiservice Requests delete gcp:stack-gcp system:serviceaccount:kube-system:generic-garbage-collector {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:kube-system:generic-garbage-collector"},"authorizationInfo":[{"granted":true,"permission":"io.crossplane.stacks.v1alpha1.stacks.delete","resource":"stacks.crossplane.io/v1alpha1/n… Kubernetes Apiservice Requests delete gcp:stack-gcp system:serviceaccount:kube-system:generic-garbage-collector 
A 2019-10-06T20:55:38.964501Z Kubernetes Apiservice Requests delete app-project1-dev:sample-stack-wordpress system:serviceaccount:kube-system:generic-garbage-collector {"@type":"type.googleapis.com/google.cloud.audit.AuditLog","authenticationInfo":{"principalEmail":"system:serviceaccount:kube-system:generic-garbage-collector"},"authorizationInfo":[{"granted":true,"permission":"io.k8s.batch.v1.jobs.delete","resource":"batch/v1/namespaces/app-project1-dev/jobs/sampl… Kubernetes Apiservice Requests delete app-project1-dev:sample-stack-wordpress system:serviceaccount:kube-system:generic-garbage-collector 
```

This clearly indicates that the resources were deleted by **generic-garbage-collector**.

On another installation I checked the first resource that is deleted, **GKECluster** in **gcp** namespace, I see that its ownerReference intended to be, **KubernetesCluster** in **app-project1-dev** namespace even *namespace* field is completely missing. This corresponds to disallowed case 1, *Namespace-scoped dependents can only specify owners in the same namespace, and owners that are cluster-scoped*.

```
kind: GKECluster
  metadata:
    name: kubernetescluster-8b6c0420-0cfd-417f-b213-1881cecf097d
    namespace: gcp
    ownerReferences:
    - apiVersion: compute.crossplane.io/v1alpha1
      kind: KubernetesCluster
      name: wordpress-cluster-f211b542-a266-4890-aa91-61161e4c5c32
      uid: 8b6c0420-0cfd-417f-b213-1881cecf097d
 ...
```

CRDs coming with stacks are also example for case 2, *Cluster-scoped dependents can only specify cluster-scoped owners, but not namespace-scoped owners.*

```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: cloudsqlinstances.database.gcp.crossplane.io
  ownerReferences:
  - apiVersion: stacks.crossplane.io/v1alpha1
    kind: ClusterStackInstall
    name: stack-gcp
    uid: 5c85331d-0955-410a-b1e2-de8c50aa433b
```

**Comment by jbw976** [comment]
Early in the initial implementation of the stack manager, I also ran into a cross namespace owner ref issue that would cause garbage collection within a few minutes of installing the stack.  We fixed that by ensuring that the `StackInstall` owner of the `Stack` were always in the same namespace.

This should be the case today.  @turkenh, do you have any insight into why the `Stack` would be garbage collected if its owner `StackInstall` is in the same namespace?

**Comment by turkenh** [comment]
> @turkenh, do you have any insight into why the Stack would be garbage collected if its owner StackInstall is in the same namespace?

This is also a question that I couldn't answer together with "why we haven't run the issue before". 
I am almost sure this is not a manual action, also double checked by following the actions that my user account did from apiserver logs.

We could have better insights if we can access logs of the *kube-controller-manager*. Since this seems not possible with gke cluster, I'll think about finding a way to reproduce this on a minikube one.

**Comment by turkenh** [comment]
I could reproduce the issue on minikube by restarting kube-controller-manager or by stopping and starting minikube again. It is happening with probability around 50% and I think I could find the root cause of why the Stack itself deleted.

The root cause seems to be that there is a wrong object reference in stack install jobs. It contains the **uid** of *ClusterStackInstall* object but kind/name of *Stack* object.

Here is what happens:
1. User creates ClusterStackInstall object:
```
apiVersion: stacks.crossplane.io/v1alpha1
kind: ClusterStackInstall
metadata:
  name: stack-gcp
  namespace: gcp
  uid: cd4622f0-553d-4412-81ba-bf3501c1bfca
```
2. StackManager creates install job with owner reference uid of *ClusterStackInstall* object but mistakenly *Stack* as kind (which is not yet created):
```
apiVersion: batch/v1
kind: Job
metadata:
  name: stack-gcp
  namespace: gcp
  ownerReferences:
  - apiVersion: stacks.crossplane.io/v1alpha1
    kind: Stack
    name: stack-gcp
    uid: cd4622f0-553d-4412-81ba-bf3501c1bfca
```
3. Then StackManager creates stack object:
```
apiVersion: stacks.crossplane.io/v1alpha1
kind: Stack
metadata:
  name: stack-gcp
  namespace: gcp
  ownerReferences:
  - apiVersion: stacks.crossplane.io/v1alpha1
    kind: ClusterStackInstall
    name: stack-gcp
    uid: cd4622f0-553d-4412-81ba-bf3501c1bfca
  uid: bfc4ca42-81c9-4d61-95a0-1d44c746a7b1
```
4. When kube-controller-manager restarts (for any reason), it builds a graph which caches dependencies among objects which is used for garbage collection. 

    i. Graph builder processes stack install job first and adds a virtual node identity for its owner (Stack object with uid of ClusterStackInstall):

    ```
    I1009 07:30:59.687310       1 graph_builder.go:550] GraphBuilder process object: batch/v1/Job, namespace gcp, name stack-gcp, uid 8de1d736-cb45-4e5d-b168-1c35c5341e60, event type add
   I1009 07:30:59.687315       1 graph_builder.go:352] add virtual node.identity: [stacks.crossplane.io/v1alpha1/Stack, namespace: gcp, name: stack-gcp, uid: cd4622f0-553d-4412-81ba-bf3501c1bfca]
    ```

    ii. Garbage collector starts processing objects, reports a [UID mismatch](https://github.com/kubernetes/kubernetes/blob/release-1.15/pkg/controller/garbagecollector/garbagecollector.go#L429) and marks Stack instance for delete:

    ```
    I1009 07:31:00.528533       1 garbagecollector.go:137] Garbage collector: all resource monitors have synced. Proceeding to collect garbage
    I1009 07:31:00.528713       1 garbagecollector.go:403] processing item [stacks.crossplane.io/v1alpha1/Stack, namespace: gcp, name: stack-gcp, uid: bfc4ca42-81c9-4d61-95a0-1d44c746a7b1]
    I1009 07:31:01.342217       1 round_trippers.go:438] GET https://localhost:8443/apis/stacks.crossplane.io/v1alpha1/namespaces/gcp/stacks/stack-gcp 200 OK in 8 milliseconds
    I1009 07:31:01.343006       1 garbagecollector.go:429] UID doesn't match, item [stacks.crossplane.io/v1alpha1/Stack, namespace: gcp, name: stack-gcp, uid: cd4622f0-553d-4412-81ba-bf3501c1bfca] not found, generating a virtual delete event
    I1009 07:31:01.343065       1 graph_builder.go:550] GraphBuilder process object: /, namespace gcp, name stack-gcp, uid cd4622f0-553d-4412-81ba-bf3501c1bfca, event type delete
    ```

    iii. Garbage collector deletes the Stack instance:

    ```
    I1009 07:31:01.413406       1 garbagecollector.go:516] delete object [stacks.crossplane.io/v1alpha1/Stack, namespace: gcp, name: stack-gcp, uid: bfc4ca42-81c9-4d61-95a0-1d44c746a7b1] with propagation policy Background
    I1009 07:31:01.425367       1 round_trippers.go:438] DELETE https://localhost:8443/apis/batch/v1/namespaces/gcp/jobs/stack-gcp 200 OK in 37 milliseconds
    ```

I'll go ahead and provide a fix, however this behaviour of GC scared me a bit :)

**Comment by jbw976** [comment]
Great detective work on finding this @turkenh! Really good debugging!

There's a part of this I don't understand still.  The owner reference of the `Stack` looks OK to me, because it's using the correct UID and Kind/APIVersion to its owner `ClusterStackInstall` object.  The `Job` object definitely has the wrong Kind set, but the `Stack` looks OK.  Why would the `Stack` be garbage collected in this case?

The log message below to me sounds like the GC has noticed that the `Stack` is **already** missing and is logging a virtual delete just to match what it has observed.

```
I1009 07:31:01.343006       1 garbagecollector.go:429] UID doesn't match, item
[stacks.crossplane.io/v1alpha1/Stack, namespace: gcp, name: stack-gcp,
uid: cd4622f0-553d-4412-81ba-bf3501c1bfca] not found, generating a virtual delete event
```

Am I misreading the owner reference of the `Stack`? It looks OK to me.

**Comment by jbw976** [comment]
Ah, I think I understand this a bit better now with your explanation and leg work, thank you @turkenh! 💪 

I agree that does sound a bit concerning that a malformed owner reference can cause the object that it "might" be referring to to be deleted. Don't think I've seen that behavior before, but that does seem like what's happening here.

**Comment by muvaf** [comment]
This just happened to me and all CRDs and CRs of the my gcp stack were gone, meaning all my infrastructure 😞  As far as I understand, the discussion here did reach to conclusion that we're doing cross-namespace owner-referencing but fix didn't directly address that. So, I'll reopen this issue.

Currently, `ClusterStackInstall` is a namespaced resource and CRDs that it installs are cluster-scoped resources and they declare `ClusterStackInstall` as their owner reference. Per https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents this is _disallowed_. And per https://github.com/kubernetes/kubernetes/issues/65200 they work on making this unacceptable by the api-server and currently it leads to unreliable behavior, which includes deletion of all CRDs (hence CRs).

IMHO, we need to get rid of owner references completely. Label selectors could be a good way to identify what's installed by a specific stack and delete when a delete request has been made.

**Comment by turkenh** [comment]
@muvaf, as stated in [this comment](https://github.com/crossplaneio/crossplane/issues/895#issuecomment-539473386), the root cause of the issue here was different and fixed, however, analysis regarding **Cross-namespace owner references** is still valid and could cause similar issues like that. 
For `ClusterStackInstall` case, [this PR](https://github.com/crossplaneio/crossplane/pull/1023) could fix the issue as it is making `ClusterStackInstall` as a cluster scoped resource. However, issue will still be valid for `StackInstall` as it also creates CRDs which are cluster scoped and we need to fix this issue by not using cross-namespace owner references as you suggested.
